text/0031-integ-testing.md

```diff
@@ -0,0 +1,542 @@
+# Integration testing for CDK apps
+
+* **Original Author(s)**: [@nija-at](https://github.com/nija-at)
+* **Tracking Issue**: [#31](https://github.com/aws/aws-cdk-rfcs/issues/31)
+* **API Bar Raiser**: @{BAR_RAISER_USER}
+
+CDK app developers and construct library authors can write integration
+tests that can be run locally or in their CI/CD pipeline.
+
+## Working Backwards - README.md
+
+### `@aws-cdk/apptest`
+
+🌩️ apptest can verify that your CDK app can be deployed successfully on AWS with the desired resources.
+
+🤖 apptest can verify that your CDK app functions as expected, such as assertions against a public API endpoint.
+
+💨 run smoke test against your production stack, and trigger a rollback when they fail.
+
+🧑‍💻 Write tests in any CDK supported language, and execute them as part of your CI pipeline.
+
+📸 Capture snapshots of your CDK apps to track how they are impacted by changes.
+
+### Writing Tests
+
+Tests that use apptest can be written in the same programming language as your CDK project.
+
+Let's start with the below CDK construct that creates an artifact repository in an S3 bucket,
+and notifies of new artifacts to an SQS queue.
+
+```ts
+// repo.ts
+import { Construct } from 'constructs';
+import { Bucket, EventType, IBucket } from '@aws-cdk/aws-s3';
+import { IQueue, Queue } from '@aws-cdk/aws-sqs';
+import { SqsDestination } from '@aws-cdk/aws-s3-notifications';
+
+export class ArtifactRepo extends Construct {
+  public readonly artifacts: IBucket;
+  public readonly publishNotifs: IQueue;
+  
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    this.artifacts = new Bucket(this, 'Artifacts');
+    this.publishNotifs = new Queue(this, 'PublishNotifs');
+
+    this.artifacts.addEventNotification(EventType.OBJECT_CREATED_PUT,
+      new SqsDestination(this.publishNotifs))
+  }
+}
+```
+
+The following integ test invokes puts an object in the S3 bucket and verifies the message in the SQS queue.
+
+```ts
+// repo.apptest.ts
+import { ArtifactRepo } from './records';
+import { AwsAssertionCall, AwsAssertion, Test } from '@aws-cdk/apptest';
+
+const repo = new ArtifactRepo(app, 'RequestRecord');
+
+const repoTest = Test.forConstruct(repo);
+
+new AwsAssertion(repoTest, 'RepoPutObject', {
+  request: AwsAssertionCall.S3.putObject({ bucket: stack.artifacts, key: '...', body: '...' })
+});
+
+new AwsAssertion(repoTest, 'MessageReceived', {
+  request: AwsAssertionCall.Sqs.receiveMessage({ queue: stack.publishNotifs }),
+  returns: { Messages: [...] }
```

Comment:
Are we supporting more than exact matches in `returns`? If the messages contain id numbers, I may not know those in advance, so I might want a language to match patterns.

Here's something like what we get in `expect` for jest:

```ts
new AwsAssertion(repoTest, 'MessageReceived', {
  request: AwsAssertionCall.Sqs.receiveMessage({ queue: stack.publishNotifs }),
  returns: AwsAssertion.objectContaining({
    Messages: AwsAssertion.arrayContaining([
      AwsAssertion.objectContaining({
        foo: 'bar'
      }),
    ]),
  }),
});
```